File: arrays/two_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> twoSum(vector<int> &nums, int target);

int main()
{
  vector<int> nums = {2, 7, 11, 19};
  for (int index : twoSum(nums, 9))
  {
    cout << index << " ";
  }
  cout << endl;
}

vector<int> twoSum(vector<int> &nums, int target)
{
  unordered_map<int, int> diff_map;
  for (int i = 0; i < nums.size(); ++i)
  {
    int cur_num = nums[i];
    int diff = target - cur_num;
    if (diff_map.find(diff) != diff_map.end())
    {
      return {diff_map[diff], i};
    }
    diff_map[cur_num] = i;
  }
  return {-1, -1};
}

File: arrays/TwoSum.kt
package arrays

import java.util.Scanner

fun main(){
    val scanner = Scanner(System.`in`);
    val size = scanner.nextInt();
    val nums: IntArray = IntArray(size);
    for(i in 0 until size){
        nums[i] = scanner.nextInt();
    }
    val target = scanner.nextInt();
    println(twoSum(nums, target).joinToString());
}

/**
 * Given an array of integers and a target
 * return index of numbers in nums which adds up to target
 * nums[i] + nums[j] = target
 *
 * solution:
 * we use a Hashmap to store the difference and it's index in a hashMap
 * and we iterate the array. If diffMap contains the difference in hashMap, return the indices
 */
fun twoSum(nums: IntArray, target: Int): IntArray {
    val diffMap = HashMap<Int, Int>();
    for((i, num) in nums.withIndex()){
        val diff = target - num;
        if(diffMap.containsKey(diff)){
            return intArrayOf(diffMap[diff]!!, i);
        }
        diffMap[num] = i;
    }
    return intArrayOf(-1, -1);
}